mutable_tree.go

```diff
@@ -501,3 +501,42 @@ func (tree *MutableTree) addOrphans(orphans []*Node) {
 		tree.orphans[string(node.hash)] = node.version
 	}
 }
+
+// NewSliceAt creates a copy of the current tree at given version. The same result
+// as deleting all other versions with DeleteVersion.
+// Uses input db as the backing database.
+func (tree *MutableTree) NewSliceAt(version int64, newDb dbm.DB) ([]byte, int64, *MutableTree, error) {
+	_, _, err := tree.SaveVersion()
+	if err != nil {
+		return nil, 0, nil, cmn.NewError("Saving tree: %v", err)
+	}
+
+	iNewTree, err := tree.GetImmutable(version)
+	if err != nil {
+		return nil, 0, nil, cmn.NewError("Getting imutable tree: %v", err)
+	}
+
+	tempTree := MutableTree{
+		ImmutableTree: iNewTree,
+		lastSaved:     tree.lastSaved,         // The most recently saved tree.
+		orphans:       make(map[string]int64), // Nodes removed by changes to working tree.
+		versions:      make(map[int64]bool),   // The previous, saved versions of the tree.
+		ndb:           tree.ndb,
+	}
+	tempTree.versions[version] = true
+
+	tempTree.root.LoadAndSetNotPersisted(&tempTree)
+	tempTree.ndb = newNodeDB(newDb, tempTree.ndb.nodeCacheSize)
+	tempTree.ndb.latestVersion = version - 1
+	if err := tempTree.ndb.SaveRoot(tempTree.root, version); err != nil {
```

Comment:
`tempTree.SaveVersion` should save the root anyway shouldn't it?

Reply:
The missing key is the `root node`, key is `r` concatenated with the `version number`. This has a known key and points to the root of the tree of `data nodes` for that version that all have hashes as keys . SaveVersion seems to treat it differently and assume its already loaded onto the database.

Reply:
OK, looks like you're saving the same root with the previous version & then `tempTree.SaveVersion()` saves the same root with new version. The important thing is to make sure that if the source tree is at version N the cloned tree is still at version N, not N+1 or something.